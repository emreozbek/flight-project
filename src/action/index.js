import { ADD_FLIGHT, FETCH_FLIGHTS, SHOW_ERROR } from "../constant";

export const addFlight = data => ({
  type: ADD_FLIGHT,
  data
});
export const fetchFlights = () => ({
  type: FETCH_FLIGHTS
});
export const showError = data => ({
  type: SHOW_ERROR,
  data
});


