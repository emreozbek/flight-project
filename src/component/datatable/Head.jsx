import React from "react";
import PropTypes from "prop-types";
import {
  TableHead,
  TableRow,
  TableCell,
  TableSortLabel
} from "@material-ui/core";
import _ from "lodash";

class Head extends React.Component {
  createSortHandler = property => event => {
    this.props.onRequestSort(event, property);
  };
  render = () => {
    const { order, orderBy, fields } = this.props;
    return (
      <TableHead>
        <TableRow>
          {_.flatMap(
            fields,
            field => (
              <TableCell
                key={field.id}
                sortDirection={orderBy === field.id ? order : false}
              >
                <TableSortLabel
                  active={orderBy === field.id}
                  direction={order}
                  onClick={this.createSortHandler(field.id)}
                >
                  {field.label}
                </TableSortLabel>
              </TableCell>
            ),
            this
          )}
        </TableRow>
      </TableHead>
    );
  };
}
Head.propTypes = {
  order: PropTypes.string.isRequired,
  orderBy: PropTypes.string.isRequired,
  fields: PropTypes.array.isRequired,
  onRequestSort: PropTypes.func
};
export default Head;
