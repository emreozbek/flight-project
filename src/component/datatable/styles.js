export const styles = theme => ({
  table: {
    minWidth: 720
  },
  tableWrapper: {
    overflowX: "auto"
  },
  searchBox: {
    paddingTop: theme.spacing.unit * 2,
    paddingLeft: theme.spacing.unit * 2,
    paddingRight: theme.spacing.unit * 2,
    paddingBottom: theme.spacing.unit * 2
  }
});
