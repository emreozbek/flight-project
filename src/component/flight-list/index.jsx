import React, { Fragment } from "react";
import { connect } from "react-redux";
import DataTable from "../datatable";
import Loading from "../loading";

export function List(props) {
  return (
    <Fragment>
      <Loading />
      <DataTable
        id="flightList"
        fields={[
          { id: "flightType", label: "Flight Type" },
          { id: "departure", label: "Departure" },
          { id: "departureDate", label: "Departure Date" },
          { id: "arrival", label: "Arrival" },
          { id: "arrivalDate", label: "Arrival Date" }
        ]}
        data={props.data}
      />
    </Fragment>
  );
}
const mapStateToProps = state => ({ data: state.reducer.data });
export default connect(mapStateToProps)(List);
