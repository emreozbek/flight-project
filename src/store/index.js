import { applyMiddleware, combineReducers, createStore } from "redux";
import { routerMiddleware, routerReducer } from "react-router-redux";
import { reducer as formReducer } from "redux-form";
import createSagaMiddleware from "redux-saga";
import { createBrowserHistory } from "history";
import reducer from "../reducer";
import Saga from "../saga";

const sagaMiddleware = createSagaMiddleware();
export const history = createBrowserHistory();
export const store = createStore(
  combineReducers({
    reducer,
    routing: routerReducer,
    form: formReducer
  }),
  applyMiddleware(sagaMiddleware, routerMiddleware(history))
);
sagaMiddleware.run(Saga);
