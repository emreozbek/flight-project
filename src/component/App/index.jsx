import React, { Component } from "react";
import { Route, Switch } from "react-router-dom";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import Theme from "../../asset/theme";
import Layout from "../Layout";
import TabMenu from "../tab-menu";
import NotFound from "../not-found";
import { fetchFlights } from "../../action";
import Snackbar from "../snackbar";

export class App extends Component {
  componentDidMount() {
    this.props.fetchFlights();
  }
  render = () => (
    <Theme>
      <Layout>
        <Snackbar />
        <Switch>
          <Route path="/" exact component={TabMenu} />
          <Route path="/add" component={TabMenu} />
          <Route component={NotFound} />
        </Switch>
      </Layout>
    </Theme>
  );
}
App.propTypes = {
  fetchFlights: PropTypes.func
};

const mapStateToProps = state => ({ ...state });
const mapDispatchToProps = {
  fetchFlights
};
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App);
