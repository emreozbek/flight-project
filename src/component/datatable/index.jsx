import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import _ from "lodash";
import {
  Table,
  TableBody,
  TableCell,
  TablePagination,
  TableRow,
  withStyles,
  InputBase,
  InputAdornment
} from "@material-ui/core";
import SearchIcon from "@material-ui/icons/Search";

import DataTableHead from "./Head";
import { styles } from "./styles";
import { getLocalStorage, setLocalStorage } from "../../utility/storage";

export class DataTable extends Component {
  defaults = {
    order: "asc",
    orderBy: this.props.fields[0].id,
    page: 0,
    rowsPerPage: 5,
    rowsPerPageOptions: [5, 10, 25, 50],
    search: ""
  };
  state = _.assign(
    {},
    this.defaults,
    this.props.options,
    JSON.parse(getLocalStorage(this.props.id))
  );
  setStorage() {
    setLocalStorage(this.props.id, JSON.stringify(this.state));
  }
  handleChangePage = (event, page) => {
    this.setState({ page }, this.setStorage);
  };
  handleChangeRowsPerPage = event => {
    this.setState({ rowsPerPage: event.target.value }, this.setStorage);
  };
  handleRequestSort = (event, orderBy) => {
    let order = "desc";
    if (this.state.orderBy === orderBy && this.state.order === "desc") {
      order = "asc";
    }
    this.setState({ order, orderBy }, this.setStorage);
  };
  filter(data, text) {
    return _.filter(data, row => {
      let isContain = false;
      _.forEach(row, value => {
        if (_.includes(_.toUpper(value), _.toUpper(text))) {
          isContain = true;
          return false;
        }
      });
      return isContain;
    });
  }
  render() {
    const { fields, classes } = this.props;
    const {
      order,
      orderBy,
      rowsPerPage,
      page,
      rowsPerPageOptions,
      search
    } = this.state;
    const data = this.filter(this.props.data, search);

    return (
      <Fragment>
        <div className={classes.searchBox}>
          <InputBase
            className={classes.input}
            placeholder="Search in the list"
            defaultValue={search}
            onChange={event =>
              this.setState({ search: event.target.value }, this.setStorage)
            }
            endAdornment={
              <InputAdornment position="end">
                <SearchIcon />
              </InputAdornment>
            }
            fullWidth
          />
        </div>

        <div className={classes.tableWrapper}>
          <Table aria-labelledby="tableTitle" className={classes.table}>
            <DataTableHead
              fields={fields}
              order={order}
              orderBy={orderBy}
              onRequestSort={this.handleRequestSort}
            />
            <TableBody>
              {_.map(
                _.slice(
                  _.orderBy(data, [orderBy], [order]),
                  page * rowsPerPage,
                  page * rowsPerPage + rowsPerPage
                ),
                n => {
                  return (
                    <TableRow tabIndex={-1} key={n.id}>
                      {_.flatMap(fields, (field, index) => (
                        <TableCell key={index}>{n[field.id]}</TableCell>
                      ))}
                    </TableRow>
                  );
                }
              )}
            </TableBody>
          </Table>
        </div>
        {data.length > 0 && (
          <TablePagination
            rowsPerPageOptions={rowsPerPageOptions}
            component="div"
            count={data.length}
            rowsPerPage={rowsPerPage}
            page={page}
            backIconButtonProps={{
              "aria-label": "Previous Page"
            }}
            nextIconButtonProps={{
              "aria-label": "Next Page"
            }}
            onChangePage={this.handleChangePage}
            onChangeRowsPerPage={this.handleChangeRowsPerPage}
          />
        )}
      </Fragment>
    );
  }
}

DataTable.propTypes = {
  id: PropTypes.string.isRequired,
  fields: PropTypes.array.isRequired,
  data: PropTypes.array.isRequired,
  options: PropTypes.object,
  classes: PropTypes.object
};

export default withStyles(styles)(DataTable);
