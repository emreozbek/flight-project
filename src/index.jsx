import React from "react";
import { render } from "react-dom";
import { Provider } from "react-redux";
import { Router } from "react-router";
import App from "./component/App";
import { store, history } from "./store";

render(
  <Provider store={store}>
    <Router history={history}>
      <App />
    </Router>
  </Provider>,
  document.getElementById("app")
);
