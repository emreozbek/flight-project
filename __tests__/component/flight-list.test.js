import React from "react";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { store } from "../../src/store";
import { List } from "../../src/component/flight-list";
import { data } from "../__mocks__/datatable";

describe("List", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <List data={data} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
