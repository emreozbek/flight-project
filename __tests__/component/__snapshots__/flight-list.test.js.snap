// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders correctly 1`] = `
Array [
  <div
    className="overlay"
  >
    <div
      className="loading-animation"
    >
      <div
        className="lds-grid"
      >
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
      </div>
    </div>
  </div>,
  <div
    className="DataTable-searchBox-3"
  >
    <div
      className="MuiInputBase-root-4 MuiInputBase-fullWidth-13 MuiInputBase-adornedEnd-9"
      onClick={[Function]}
    >
      <input
        className="MuiInputBase-input-14 MuiInputBase-inputAdornedEnd-20"
        defaultValue=""
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Search in the list"
        type="text"
      />
      <div
        className="MuiInputAdornment-root-21 MuiInputAdornment-positionEnd-24"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-26"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </div>
    </div>
  </div>,
  <div
    className="DataTable-tableWrapper-2"
  >
    <table
      aria-labelledby="tableTitle"
      className="MuiTable-root-35 DataTable-table-1"
    >
      <thead
        className="MuiTableHead-root-36"
      >
        <tr
          className="MuiTableRow-root-37 MuiTableRow-head-40"
        >
          <th
            aria-sort="ascending"
            className="MuiTableCell-root-42 MuiTableCell-head-43"
            scope="col"
          >
            <span
              className="MuiButtonBase-root-59 MuiTableSortLabel-root-54 MuiTableSortLabel-active-55"
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="button"
              tabIndex="0"
            >
              Flight Type
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-26 MuiTableSortLabel-icon-56 MuiTableSortLabel-iconDirectionAsc-58"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-head-43"
            scope="col"
          >
            <span
              className="MuiButtonBase-root-59 MuiTableSortLabel-root-54"
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="button"
              tabIndex="0"
            >
              Departure
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-26 MuiTableSortLabel-icon-56 MuiTableSortLabel-iconDirectionAsc-58"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-head-43"
            scope="col"
          >
            <span
              className="MuiButtonBase-root-59 MuiTableSortLabel-root-54"
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="button"
              tabIndex="0"
            >
              Departure Date
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-26 MuiTableSortLabel-icon-56 MuiTableSortLabel-iconDirectionAsc-58"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-head-43"
            scope="col"
          >
            <span
              className="MuiButtonBase-root-59 MuiTableSortLabel-root-54"
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="button"
              tabIndex="0"
            >
              Arrival
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-26 MuiTableSortLabel-icon-56 MuiTableSortLabel-iconDirectionAsc-58"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-head-43"
            scope="col"
          >
            <span
              className="MuiButtonBase-root-59 MuiTableSortLabel-root-54"
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="button"
              tabIndex="0"
            >
              Arrival Date
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-26 MuiTableSortLabel-icon-56 MuiTableSortLabel-iconDirectionAsc-58"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root-62"
      >
        <tr
          className="MuiTableRow-root-37"
          tabIndex={-1}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-body-44"
          />
          <td
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-body-44"
          />
          <td
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-body-44"
          />
          <td
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-body-44"
          />
          <td
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-body-44"
          />
        </tr>
        <tr
          className="MuiTableRow-root-37"
          tabIndex={-1}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-body-44"
          />
          <td
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-body-44"
          />
          <td
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-body-44"
          />
          <td
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-body-44"
          />
          <td
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-body-44"
          />
        </tr>
        <tr
          className="MuiTableRow-root-37"
          tabIndex={-1}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-body-44"
          />
          <td
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-body-44"
          />
          <td
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-body-44"
          />
          <td
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-body-44"
          />
          <td
            aria-sort={null}
            className="MuiTableCell-root-42 MuiTableCell-body-44"
          />
        </tr>
      </tbody>
    </table>
  </div>,
  <div
    className="MuiTablePagination-root-63"
  >
    <div
      className="MuiToolbar-root-73 MuiToolbar-regular-75 MuiToolbar-gutters-74 MuiTablePagination-toolbar-64"
    >
      <div
        className="MuiTablePagination-spacer-65"
      />
      <span
        className="MuiTypography-root-77 MuiTypography-caption-87 MuiTypography-colorInherit-106 MuiTablePagination-caption-66"
      >
        Rows per page:
      </span>
      <div
        className="MuiInputBase-root-4 MuiTablePagination-input-70"
        onClick={[Function]}
      >
        <div
          className="MuiSelect-root-113 MuiTablePagination-selectRoot-67"
        >
          <div
            aria-haspopup="true"
            aria-pressed="false"
            className="MuiSelect-select-114 MuiTablePagination-select-68 MuiSelect-selectMenu-117 MuiInputBase-input-14"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            5
          </div>
          <input
            type="hidden"
            value={5}
          />
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-26 MuiSelect-icon-119 MuiTablePagination-selectIcon-69"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
      <span
        className="MuiTypography-root-77 MuiTypography-caption-87 MuiTypography-colorInherit-106 MuiTablePagination-caption-66"
      >
        1-3 of 3
      </span>
      <div
        className="MuiTablePagination-actions-72"
      >
        <button
          aria-label="Previous Page"
          className="MuiButtonBase-root-59 MuiButtonBase-disabled-60 MuiIconButton-root-124 MuiIconButton-colorInherit-125 MuiIconButton-disabled-128"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="-1"
          type="button"
        >
          <span
            className="MuiIconButton-label-129"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-26"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Next Page"
          className="MuiButtonBase-root-59 MuiButtonBase-disabled-60 MuiIconButton-root-124 MuiIconButton-colorInherit-125 MuiIconButton-disabled-128"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="-1"
          type="button"
        >
          <span
            className="MuiIconButton-label-129"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-26"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>,
]
`;
